aggregator/handler/common/traffic_controller.go

```diff
@@ -0,0 +1,184 @@
+
+	// InstrumentOptions returns the instrument options.
+	InstrumentOptions() instrument.Options
+}
+
+type trafficControlOptions struct {
+	store             kv.Store
+	defaultDisabled   bool
+	runtimeDisableKey string
+	initTimeout       time.Duration
+	instrumentOpts    instrument.Options
+}
+
+// NewTrafficControlOptions creats a new TrafficControlOptions.
+func NewTrafficControlOptions() TrafficControlOptions {
+	return &trafficControlOptions{
+		defaultDisabled: defaultDefaultDisabled,
+		initTimeout:     defaultInitTimeout,
+		instrumentOpts:  instrument.NewOptions(),
+	}
+}
+
+func (opts *trafficControlOptions) SetStore(store kv.Store) TrafficControlOptions {
+	o := *opts
+	o.store = store
+	return &o
+}
+
+func (opts *trafficControlOptions) Store() kv.Store {
+	return opts.store
+}
+
+func (opts *trafficControlOptions) SetDefaultDisabled(value bool) TrafficControlOptions {
+	o := *opts
+	o.defaultDisabled = value
+	return &o
+}
+
+func (opts *trafficControlOptions) DefaultDisabled() bool {
+	return opts.defaultDisabled
+}
+
+func (opts *trafficControlOptions) SetRuntimeDisableKey(value string) TrafficControlOptions {
+	o := *opts
+	o.runtimeDisableKey = value
+	return &o
+}
+
+func (opts *trafficControlOptions) RuntimeDisableKey() string {
+	return opts.runtimeDisableKey
+}
+
+func (opts *trafficControlOptions) SetInitTimeout(value time.Duration) TrafficControlOptions {
+	o := *opts
+	o.initTimeout = value
+	return &o
+}
+
+func (opts *trafficControlOptions) InitTimeout() time.Duration {
+	return opts.initTimeout
+}
+
+func (opts *trafficControlOptions) SetInstrumentOptions(value instrument.Options) TrafficControlOptions {
+	o := *opts
+	o.instrumentOpts = value
+	return &o
+}
+
+func (opts *trafficControlOptions) InstrumentOptions() instrument.Options {
+	return opts.instrumentOpts
+}
+
+// TrafficController controls if traffic is enabled.
+type TrafficController struct {
+	disabled *atomic.Bool
+}
+
+// NewTrafficController creates a new traffic controller.
+func NewTrafficController(opts TrafficControlOptions) *TrafficController {
+	disabled := atomic.NewBool(opts.DefaultDisabled())
+	iOpts := opts.InstrumentOptions()
+	newUpdatableFn := func() (watch.Updatable, error) {
+		w, err := opts.Store().Watch(opts.RuntimeDisableKey())
+		return w, err
+	}
+	getFn := func(updatable watch.Updatable) (interface{}, error) {
+		return updatable.(kv.ValueWatch).Get(), nil
+	}
+	processFn := func(update interface{}) error {
+		b, err := util.BoolFromValue(
+			update.(kv.Value),
+			opts.RuntimeDisableKey(),
+			opts.DefaultDisabled(),
+			util.NewOptions().SetLogger(iOpts.Logger()),
+		)
+		if err != nil {
+			return err
+		}
+		disabled.Store(b)
+		return nil
+	}
+	vOptions := watch.NewOptions().
+		SetInitWatchTimeout(opts.InitTimeout()).
+		SetInstrumentOptions(iOpts).
+		SetNewUpdatableFn(newUpdatableFn).
+		SetGetUpdateFn(getFn).
+		SetProcessFn(processFn)
+	value := watch.NewValue(vOptions)
+	value.Watch()
```

Comment:
Probably should have a `Close` method on the traffic controller as well to unwatch the runtime value or otherwise this leaks goroutines.